core/templates/pages/contributor-dashboard-admin-page/services/format-contribution-attributes.service.ts

```diff
@@ -0,0 +1,256 @@
+    });
+    this.contributionAttribute.push({
+      key: 'Active Topics',
+      value: this.getTextForActiveTopicContributionAttribute(
+        element.topicsWithTranslationSubmissions)
+    });
+  }
+
+  private getQuestionSubmitterContributionAttributes(
+      element: QuestionSubmitterStats): void {
+    this.contributionAttribute.push({
+      key: 'Submitted Questions',
+      value: this.getTextForContributionAttribute(
+        element.submittedQuestionsCount)
+    });
+    this.contributionAttribute.push({
+      key: 'Accepted Questions',
+      value: this.getAcceptedSubmittedTextForContributionAttribute(
+        element.acceptedQuestionsCount,
+        element.acceptedQuestionsWithoutReviewerEditsCount)
+    });
+    this.contributionAttribute.push({
+      key: 'Rejected Questions',
+      value: this.getTextForContributionAttribute(
+        element.rejectedQuestionsCount)
+    });
+    this.contributionAttribute.push({
+      key: 'Active Topics',
+      value: this.getTextForActiveTopicContributionAttribute(
+        element.topicsWithQuestionSubmissions)
+    });
+  }
+
+  private getQuestionReviewerContributionAttributes(
+      element: QuestionReviewerStats): void {
+    this.contributionAttribute.push({
+      key: 'Reviewed Questions',
+      value: element.reviewedQuestionsCount.toString()
+    });
+    this.contributionAttribute.push({
+      key: 'Accepted Questions',
+      value: this.getAcceptedReviewedTextForContributionAttribute(
+        element.acceptedQuestionsCount,
+        element.acceptedQuestionsWithReviewerEditsCount)
+    });
+    this.contributionAttribute.push({
+      key: 'Rejected Questions',
+      value: this.getTextForContributionAttribute(
+        element.rejectedQuestionsCount)
+    });
+    this.contributionAttribute.push({
+      key: 'Active Topics',
+      value: this.getTextForActiveTopicContributionAttribute(
+        element.topicsWithQuestionReviews)
+    });
+  }
+
+  /**
+  * Return an array of attributes of a contribution in the correct
+  * format to appear in the contribution admin stats component
+  * expanded detail section
+  *
+  * @param {TranslationSubmitterStats |
+  *         TranslationReviewerStats |
+  *         QuestionSubmitterStats |
+  *         QuestionReviewerStats} element - contribution data, the type
+  * differs depending on the contribution type and sub type
+  * @returns {ContributionAttribute[]} - An array of
+  * contribution attributes, with the value correctly formatted for display
+  */
+  getContributionAttributes(
+      element: TranslationSubmitterStats |
+               TranslationReviewerStats |
+               QuestionSubmitterStats |
+               QuestionReviewerStats):
+      ContributionAttribute[] {
+    this.contributionAttribute = [
+      {
+        key: 'Date Joined',
+        value: element.firstContributionDate
+      }
+    ];
+
+    if (element instanceof TranslationSubmitterStats) {
```

Comment:
The way four different things are handled here is a bit messy type-wise, I think. How about this: instead, on the TranslationSubmitterStats object (ditto for others), define a method saying .getContributionAttributes(), then here you can do element.getContributionAttributes() generically in the caller.  

Those individual methods can call common functions in contribution-attributes-formatter.service.ts as needed.

Reply:
Done